### PR TITLE
S✔ ◾ Fix #808: Desktop shaves missing embedUrl/videoFile (backend backstop)

### LINKED ISSUE #808: 🐛 Bug - Desktop-recorded YakShaves sometimes missing embedUrl/videoFile causing no preview
Cc: @calumjs @ricksu978 @adamcogan
Hi YakShaver Desktop team

[▶️ Watch the video (01:52)](https://www.youtube.com/watch?v=ArwY4D8VkRs)

### Describe the Bug
In Tenant view, some YakShaves created from the Desktop app intermittently show no video preview. Debugging indicates that for some Desktop-recorded items the backend payload is missing the videoFile and/or embedUrl, while other Desktop-recorded items correctly include embedUrl and render a preview.

### To Reproduce
1. Open YakShaver Desktop app
2. Record a YakShave using the Desktop source
3. Open the item in the Tenant view
4. See error (no video preview shown; related videoFile/embedUrl missing in debug JSON)

### Expected Behavior
YakShaves recorded from the Desktop source should always include the required video metadata (videoFile and embedUrl) so the Tenant view can render a preview consistently.

### Tasks
- [ ] Investigate
- [ ] Fix

### More Information
Tenant view showing YakShaves without previews

Observed inconsistency:
- Some Desktop-recorded YakShaves have embedUrl and preview works.
- Some Desktop-recorded YakShaves have missing embedUrl/videoFile so preview does not render.

### Environment
 - Device: Windows PC
 - Browser: N/A
 - OS: Windows

### Screenshots

Thanks!

### PR DESCRIPTION
**Pull request by calumjs**
Fixes #808.

## Root cause
#808 reports the symptom in the **Tenant view** (the portal/web app): YakShaves recorded from the Desktop source are missing the `videoFile`/`embedUrl`, so the Tenant view renders no preview.

The Tenant-view preview is driven **exclusively by the portal `WorkItemDto`'s `uploadedVideo*` fields** (`uploadedVideoProvider` / `uploadedVideoEmbedUrl` / `uploadedVideoUrl`). Those fields were previously left to the **LLM** to copy out of the system prompt during structured extraction (`orchestrator.convertToObjectAsync(mcpResult, WorkItemDtoSchema)`) — a non-deterministic step that intermittently dropped them. That is the exact "missing embedUrl/videoFile" symptom #808 reports.

(Separately, the desktop app's own thumbnail list — `VideoThumbnail.tsx` / `ShaveCards.tsx` — reads the **local** `videoEmbedUrl` DB column, which was written only on the renderer side in reaction to a race-prone `uploading_video`/`downloading_video` progress event. That column is desktop-only and is **never** sent to the portal, so it cannot affect the Tenant view — see the backstop note below.)

## Fix — the Tenant-view payload (this is what resolves #808)
Set the portal `WorkItemDto`'s `uploadedVideo*` fields **deterministically** from the authoritative upload result, overriding whatever the model produced, before `SendWorkItemDetailsToPortal` (`src/backend/ipc/process-video-handlers.ts`). A successful recording now ALWAYS carries the embed URL to the portal regardless of model output.

The derivation + override are extracted into pure, unit-testable helpers (`src/backend/services/workflow/video-metadata-persistence.ts`):
- `derivePortalVideoFields(result)` — derives `{ provider, url, embedUrl }`, synthesizing an **iframe-embeddable** form for recognized providers (YouTube `/embed/`, Vimeo `player.vimeo.com/video/<id>[?h=]`), and leaving `embedUrl` null for unrecognized providers so the Tenant view falls back to a link rather than a broken `<iframe>` src. Returns `null` (leave model output untouched) when there's no usable URL.
- `applyPortalVideoFields(dto, result)` — applies that derivation onto the `WorkItemDto` (mirrors the local-DB `applyVideoMetadataPersistence` helper), so the override wiring is unit-testable: it fires before the portal POST, overwrites only the three video fields, and is skipped on a null derivation.

## Secondary fix — local-DB backstop (desktop-thumbnail robustness, not the Tenant view)
A **separate** improvement, independent of #808's Tenant-view symptom: a backend backstop in `processVideoSource` persists the **local** `videoEmbedUrl` column / attaches the video source directly from the authoritative result, so the desktop app's own thumbnail list is robust to the renderer-side progress-event race. This column is desktop-only and is **never** transmitted to the portal, so it does not (and cannot) change the Tenant-view preview — it is included here because it shares the same authoritative result and fixes a real, adjacent desktop-side gap.

- **Upload origin** → set local `videoEmbedUrl`, only when the shave doesn't already have one (never clobbers a UI/metadata-update write).
- **External origin** → attach a video source via the idempotent `attachVideoSourceToShave`.
- No-op when the upload didn't succeed or produced no URL. Best-effort / non-fatal.

## Tests
`src/backend/services/workflow/video-metadata-persistence.test.ts` — DB-free unit tests:
- **Tenant-view payload** (the #808 fix): `derivePortalVideoFields` provider/url/embed derivation (YouTube, `youtu.be`, Vimeo page + unlisted hash, unrecognized provider, failed/no-URL); `applyPortalVideoFields` override wiring (overwrites the three fields, leaves other dto fields untouched, skips on null derivation, clears a model-guessed embed URL to null for unrecognized providers).
- **Local backstop** (desktop-thumbnail robustness): `decideVideoMetadataPersistence` + `applyVideoMetadataPersistence` wiring against an in-memory store (writes when missing, no-clobber, external attach, no-op branches).

## Verification
- `npm run build` (tsc): green
- `npx vitest run --exclude '**/db/**'` : 187 passed (DB-backed suites excluded — better-sqlite3 native ABI mismatch in this env; ffmpeg native binary unsupported on win32-arm64, both pre-existing and unrelated)
- `npx biome check` on changed files: clean

## Pending — live walkthrough (required to close #808)
#808's acceptance criterion is **behavioural and Tenant-side**: a real Desktop recording must land in the portal and render a preview in the Tenant view consistently. That portal → Tenant render path is **not** demonstrated by the unit tests, which prove only the deterministic derivation/override and the local-DB wiring. Per repo house rules this needs a **live repro-then-fix walkthrough** demonstrating the AC: record via Desktop on a remote project, confirm the portal payload carries `uploadedVideoEmbedUrl`, and confirm the Tenant view renders the preview. Tracking that before considering #808 closed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by calumjs — COMMENTED** [review]
## Review summary

This PR adds a backstop that persists `videoEmbedUrl` to the local SQLite shave record, plus 7 unit tests over the pure decision function. The decision-layer logic is sound, but there are two structural problems: the fix targets a different store than issue #808 reports, and the verification only exercises the decision branch — not the persistence path the fix exists to guarantee.

---

### High — Fix targets the local DB, not the Tenant-view payload that #808 reports

Issue #808 is scoped to the **Tenant view** (portal web UI): *"the backend payload is missing the videoFile and/or embedUrl."* The Tenant view is fed by the **portal** `WorkItemDto` POSTed to `/desktopapp/post-task-record`, whose video fields (`uploadedVideoProvider`, `uploadedVideoEmbedUrl`, `uploadedVideoUrl`) live at `src/backend/services/portal/actions.ts:79-81`.

That DTO is built by an LLM call: `orchestrator.convertToObjectAsync(mcpResult, WorkItemDtoSchema)` (`src/backend/ipc/process-video-handlers.ts:619-623`), which runs `generateObject` over `mcpResult`. The embed URL only reaches `mcpResult` via system-prompt injection (`src/backend/services/mcp/mcp-orchestrator.ts:743-752`, *"Embed this URL in the task content"*). After extraction, only `projectId`/`projectName` are overridden deterministically (lines 624-625) — the video fields are left exactly as the model produced them. That model-derived population is precisely why the bug is intermittent.

This PR's backstop (`persistVideoMetadataToShave` → `shaveService.updateShave(shaveId, { videoEmbedUrl })`) writes **only** to the local `shaves.videoEmbedUrl` column — a different store. The portal sync write-back (`process-video-handlers.ts:628-636`) flows portal→local and omits `videoEmbedUrl`, so the local value never reaches the portal. The surface this PR actually fixes is the desktop *My Shaves* thumbnail (`src/ui/src/components/shaves/VideoThumbnail.tsx:10` reads `shave.videoEmbedUrl`), not the Tenant view #808 screenshots.

**Remedy:** Deterministically set `workItemDto.uploadedVideoEmbedUrl` / `uploadedVideoUrl` (and provider) from `youtubeResult.data.url` before `SendWorkItemDetailsToPortal(workItemDto)`, instead of relying on the LLM to copy the URL out of the system prompt. If the desktop *My Shaves* preview is also in scope, scope the PR/issue explicitly to that surface and note the Tenant-view path remains open. Worth confirming with the reporter which surface is the real target.

---

### Medium — One-sided verification: the persistence path is never exercised

All 7 tests in `src/backend/services/workflow/video-metadata-persistence.test.ts` assert only on the **return value** of the pure `decideVideoMetadataPersistence` (e.g. `expect(action).toEqual({kind:'setEmbedUrl',...})`). They import only vitest, the `VideoUploadResult` type, and the function under test — no `ShaveService`, no `ProcessVideoIPCHandlers`. There is no test for `process-video-handlers.ts` at all.

So the actual backstop — `persistVideoMetadataToShave` → `switch(action.kind)` → `ShaveService.getShaveById`/`updateShave`/`attachVideoSourceToShave` → DB — is exercised by **zero** tests. The decision that *precedes* the write is proven; the write itself (and the no-clobber skip) is not. The PR body confirms this: *"Live repro / walkthrough pending — verification was code + build + unit-test only."*

**Remedy:**
1. Add a wiring test for `persistVideoMetadataToShave` with a fake/in-memory `ShaveService`, asserting `updateShave({ videoEmbedUrl })` **is** called when the shave has none and **is not** called when it already has one (proves both the write and the no-clobber branch). The three `ShaveService` methods already exist in `src/backend/services/shave/shave-service.ts`, so this is straightforward.
2. Complete the pending live happy-path verification: a real recording confirming `videoEmbedUrl` is persisted and the preview renders.

File: src/backend/services/workflow/video-metadata-persistence.test.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
+import { describe, expect, it } from "vitest";
```
Comment (calumjs) [inline src/backend/services/workflow/video-metadata-persistence.test.ts]: All 7 tests here assert only on the return of the pure `decideVideoMetadataPersistence`; none import `ShaveService` or `ProcessVideoIPCHandlers`. The persistence path the fix exists to guarantee (`persistVideoMetadataToShave` → `ShaveService.updateShave`/`attachVideoSourceToShave` → DB) is exercised by zero tests. Add a wiring test with a fake `ShaveService` asserting `updateShave({videoEmbedUrl})` is called when the shave has none and skipped when it already has one — to prove the write and the no-clobber branch, not just the decision.

File: src/backend/ipc/process-video-handlers.ts [hunk 2]
```diff
@@ -437,6 +438,15 @@ export class ProcessVideoIPCHandlers {
...
       this.lastVideoFilePath = filePath;
 
+      // #808: Persist the video embed URL / source onto the shave record directly from the
+      // authoritative upload result. The UI normally writes this in response to a workflow
+      // progress event, but that event can be missed or coalesced (e.g. the workflow advances
+      // past `uploading_video` before the renderer subscribes, or the renderer's in-memory
+      // dedup set drops it), leaving the saved shave without `videoEmbedUrl`/`videoFile` and so
+      // with no preview in the Tenant view. Writing it here from the backend guarantees the
+      // field is persisted whenever the upload/download succeeded, regardless of UI timing.
+      this.persistVideoMetadataToShave(shaveId, youtubeResult);
```
Comment (calumjs) [inline src/backend/ipc/process-video-handlers.ts:451]: #808 reports the **Tenant view** (portal payload) missing the embed URL — but this backstop writes only to the local `shaves.videoEmbedUrl` column. The portal `WorkItemDto` POSTed in the same flow (built ~line 619-623 via `convertToObjectAsync` → `generateObject`) populates its video fields (`uploadedVideoEmbedUrl`/`uploadedVideoUrl`, `portal/actions.ts:79-81`) from the LLM, and only `projectId`/`projectName` are overridden deterministically afterwards. The local write never reaches that payload (the portal sync flows portal→local and omits `videoEmbedUrl`). To fix the reported Tenant-view symptom, also set `workItemDto.uploadedVideoEmbedUrl`/`uploadedVideoUrl` deterministically from `youtubeResult.data.url` before `SendWorkItemDetailsToPortal`. As-is this fixes the desktop *My Shaves* thumbnail (`VideoThumbnail.tsx:10`), a different surface.

File: src/backend/services/workflow/video-metadata-persistence.test.ts (hunk 1 above)
Comment (calumjs) [reply src/backend/services/workflow/video-metadata-persistence.test.ts]: Fixed in 286337d. Extracted the action→ShaveService wiring out of the IPC handler into `applyVideoMetadataPersistence(store, shaveId, youtubeResult)` (takes a minimal `VideoMetadataShaveStore` interface), and added wiring tests against an in-memory fake that assert:

- `updateShave({ videoEmbedUrl })` is called exactly once when the shave has no embed URL,
- it is NOT called (action `none`, no clobber) when the shave already has one,
- the external branch calls `attachVideoSourceToShave` instead of `updateShave`,
- and the no-shave-id / shave-not-found cases write nothing.

`persistVideoMetadataToShave` now just delegates to that function, so the persisted write path is proven, not only the decision.

File: src/backend/ipc/process-video-handlers.ts (hunk 2 above)
Comment (calumjs) [reply src/backend/ipc/process-video-handlers.ts:451]: Fixed in 286337d. You're right — the local-column write was the My Shaves thumbnail surface, not the Tenant-view payload #808 reports. The portal `WorkItemDto` video fields were being left to the LLM (`convertToObjectAsync` → `generateObject`), which is the non-deterministic, intermittently-dropped path.

I now derive them deterministically from the authoritative `youtubeResult` via `derivePortalVideoFields(...)` and override the model output before `SendWorkItemDetailsToPortal`:

```ts
const portalVideoFields = derivePortalVideoFields(youtubeResult);
if (portalVideoFields) {
  workItemDto.uploadedVideoProvider = portalVideoFields.uploadedVideoProvider;
  workItemDto.uploadedVideoEmbedUrl = portalVideoFields.uploadedVideoEmbedUrl;
  workItemDto.uploadedVideoUrl = portalVideoFields.uploadedVideoUrl;
}
```

`derivePortalVideoFields` normalizes the URL: for YouTube it sets provider `youtube`, the canonical `watch?v=ID` URL, and the `/embed/ID` form; for a non-YouTube URL it mirrors the raw URL with a null provider; and it returns `null` on a failed/empty result so we never overwrite model output with nulls. Covered by new unit tests (YouTube watch + youtu.be short links, non-YouTube, failed, no-URL). So a successful recording now always carries the embed URL into the portal payload regardless of model output.

**Review by calumjs — COMMENTED** [review]
## Review summary

One confirmed finding from this pass, grouped by severity below.

### Medium

**Non-YouTube (external) uploads populate `uploadedVideoEmbedUrl` with a non-embeddable page URL**

In `src/backend/services/workflow/video-metadata-persistence.ts`, `derivePortalVideoFields`'s non-YouTube branch returns `{ uploadedVideoEmbedUrl: url }` — the raw page URL — for any source with `origin: "external"`. The YouTube branch correctly synthesizes the embeddable `https://www.youtube.com/embed/<id>` form, but the external branch does not.

`uploadedVideoEmbedUrl` is the *embed* field (distinct from `uploadedVideoUrl`) and the portal's Tenant view consumes it as an `<iframe>` `src`. A plain page URL like `https://vimeo.com/123456` is **not** iframe-embeddable — Vimeo's embeddable form is `https://player.vimeo.com/video/<id>` (and unlisted videos additionally need `?h=<hash>`). So on the external path the very field this fix exists to populate carries a value the iframe cannot load, and the preview still won't render.

This path is reachable with real input: `IPC_CHANNELS.PROCESS_VIDEO_URL` accepts an arbitrary user-supplied URL -> `processUrlVideo` -> `youtubeDownloadService.getVideoMetadata`, which runs the URL through yt-dlp. yt-dlp supports Vimeo and ~1800 other sites, so a Vimeo URL genuinely returns `origin: "external"` with `webpage_url = https://vimeo.com/<id>`, which then flows into `derivePortalVideoFields` and onto the portal DTO.

The accompanying test enshrines the wrong value — it asserts `derivePortalVideoFields` for `https://vimeo.com/123456` yields `uploadedVideoEmbedUrl: 'https://vimeo.com/123456'` — so the gap is baked into the suite rather than caught.

Evidence (Vimeo): `player.vimeo.com/video/<id>` is the embeddable iframe form; plain `vimeo.com` URLs are page URLs that do not work for embedding or player parameters — https://help.vimeo.com/hc/en-us/articles/30021987305873-How-do-I-embed-parameters-to-a-Vimeo-com-URL and https://developer.vimeo.com/api/oembed/videos

Remedy — pick one:
- Resolve the embeddable form for known non-YouTube providers (e.g. Vimeo via its oEmbed endpoint / `player.vimeo.com/video/<id>`) before assigning `uploadedVideoEmbedUrl`; or
- Set `uploadedVideoEmbedUrl` to `null` for unrecognized providers (carry only `uploadedVideoUrl`) so the Tenant view falls back to a link instead of a broken iframe; and update the test to assert that contract; or
- At minimum, since the desktop happy path uploads to YouTube, explicitly scope/document that external non-YouTube sources won't get an embeddable preview.

Note: the desktop-local UI (`UploadResult.tsx`) renders external videos as a link, not an iframe, so the desktop UI itself is unaffected — the impact is on the portal consumer.

File: src/backend/services/workflow/video-metadata-persistence.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      uploadedVideoEmbedUrl: `https://www.youtube.com/embed/${youtubeId}`,
+    };
+  }
+
+  // Non-YouTube URL: still carry it deterministically, but we can't synthesize an embed form
+  // or claim a provider, so mirror the raw URL and leave the provider unknown.
+  return {
+    uploadedVideoProvider: null,
+    uploadedVideoUrl: url,
+    uploadedVideoEmbedUrl: url,
```
Comment (calumjs) [inline src/backend/services/workflow/video-metadata-persistence.ts]: Medium: for the non-YouTube (`origin: "external"`) path this assigns the raw page URL to `uploadedVideoEmbedUrl`, but the portal's Tenant view consumes that field as an `<iframe>` `src`. A plain `https://vimeo.com/<id>` page URL is not iframe-embeddable (Vimeo's embeddable form is `https://player.vimeo.com/video/<id>`, and unlisted videos additionally need `?h=<hash>`), so the preview this fix exists to enable still won't render for external sources. yt-dlp reaches this branch for Vimeo + ~1800 other sites via the user-supplied URL path. Either resolve the embeddable form for known providers (e.g. Vimeo oEmbed -> `player.vimeo.com/video/<id>`), or set `uploadedVideoEmbedUrl` to `null` for unrecognized providers so the consumer falls back to a link instead of a broken iframe. The matching test currently enshrines the page URL as the embed URL rather than catching this.

Comment (calumjs) [reply src/backend/services/workflow/video-metadata-persistence.ts]: Fixed in 58d86e3. `derivePortalVideoFields` now only ever puts an iframe-embeddable URL in `uploadedVideoEmbedUrl`:

- **Vimeo** page URLs are resolved to the player form — `https://vimeo.com/123456` -> `https://player.vimeo.com/video/123456`, and the unlisted privacy hash is carried through as `?h=`: `https://vimeo.com/123456/ab12cd34ef` -> `https://player.vimeo.com/video/123456?h=ab12cd34ef`. Provider is set to `"vimeo"`.
- **Unrecognized providers** now set `uploadedVideoEmbedUrl: null` (carrying only `uploadedVideoUrl`), so the Tenant view falls back to a link instead of rendering an `<iframe>` whose `src` would fail to load.
- YouTube `/embed/<id>` path is unchanged.

The test that enshrined the page URL as the embed URL was replaced, and I added coverage for the Vimeo page branch, the unlisted-hash branch, and the null-embed unrecognized-provider branch.

Validation: `tsc` build clean, `npx vitest run` non-DB suite green (183 passed; the only failures are pre-existing native-binary issues on this win32-arm64 box — better-sqlite3 db tests + ffmpeg-service), and biome clean on the changed files (the one remaining biome error is a pre-existing organize-imports assist in `VideoPreviewModal.tsx`, untouched here).

**Review by calumjs — COMMENTED** [review]
## Review — #808 missing embedUrl/videoFile (backend backstop)

The deterministic portal-field override (`derivePortalVideoFields` → `workItemDto.uploadedVideo*` before `SendWorkItemDetailsToPortal`) is the right shape for the actual Tenant-view symptom #808 reports. Three medium findings below, all about coverage and framing rather than code correctness.

### Medium

**1. The override that actually fixes #808 has no wiring test** — `src/backend/ipc/process-video-handlers.ts:646-651`

`derivePortalVideoFields()` is thoroughly unit-tested in isolation, but the wiring that mutates the LLM-produced `workItemDto` is not. Nothing pins that the override (a) fires *before* the portal POST, (b) leaves the model's other `workItemDto` fields intact, or (c) is skipped — leaving model output in place — when `derivePortalVideoFields` returns `null`. A future edit could drop the assignment or reorder it after `SendWorkItemDetailsToPortal` and every test would stay green.

This is also self-inconsistent within the PR: the symmetric local-metadata path *was* extracted into a testable `applyVideoMetadataPersistence(...)` helper with wiring tests that assert `updateShave` is/isn't called, while the portal path has no equivalent helper and no wiring test.

Remedy: add a handler/integration-level test, or extract a thin `applyPortalVideoFields(dto, result)` helper mirroring `applyVideoMetadataPersistence`, asserting the three fields are overridden from the result, the dto is otherwise untouched, and a `null` derivation leaves the model values in place.

**2. PR framing misattributes which change closes #808** — `src/backend/ipc/process-video-handlers.ts:441` (`persistVideoMetadataToShave`) + PR body

#808's symptom is a missing preview in the **Tenant view**, which is driven exclusively by the portal `WorkItemDto`'s `uploadedVideo*` fields. The local `videoEmbedUrl` column is read only by the desktop UI (`VideoThumbnail.tsx:10`, `ShaveCards.tsx:26`) and is never sent to the portal — the portal-sync writeback copies portal IDs/title back *in*, it does not push `videoEmbedUrl` *out*. So the change that actually resolves the reported symptom is `derivePortalVideoFields`, yet the PR body (title, Root cause, Fix, Tests, Pending) describes only the local-DB backstop and never mentions `derivePortalVideoFields` / `uploadedVideoEmbedUrl`. Roughly 5 of the 7 tests cover the desktop-thumbnail surface, not the Tenant-view surface.

The local backstop is a legitimate desktop-thumbnail robustness improvement — but as written the narrative creates a false sense that it alone closes #808.

Remedy: reframe so the portal-field derivation is identified as the #808 fix, and describe the local-DB backstop as a separate desktop-thumbnail improvement.

**3. The behavioural acceptance criterion is not demonstrated** — Verification / "Pending" section

#808's acceptance criterion is behavioural and Tenant-side: Desktop recordings should *always* carry the video metadata so the Tenant view renders a preview consistently. Verification here is build + DB-free/in-memory unit tests only; the PR states "Live repro / walkthrough pending." Nothing exercises whether the portal accepts the `uploadedVideo*` field names or whether the Tenant view renders from them. Per repo convention (live repro-then-fix walkthroughs demonstrating every AC), this AC needs a live run.

Remedy: before considering #808 closed, do a live Desktop recording on a remote project, confirm the portal payload carries `uploadedVideoEmbedUrl`, confirm the Tenant view renders the preview, and attach the walkthrough.

File: src/backend/ipc/process-video-handlers.ts [hunk 4]
```diff
@@ -624,6 +637,19 @@ export class ProcessVideoIPCHandlers {
...
+            // structured extraction, which intermittently dropped them — the exact "missing
+            // embedUrl/videoFile" symptom #808 reports. Set them deterministically from the
+            // same authoritative upload result the local backstop uses, so a successful
+            // recording ALWAYS carries the embed URL to the portal regardless of model output.
+            const portalVideoFields = derivePortalVideoFields(youtubeResult);
+            if (portalVideoFields) {
+              workItemDto.uploadedVideoProvider = portalVideoFields.uploadedVideoProvider;
+              workItemDto.uploadedVideoEmbedUrl = portalVideoFields.uploadedVideoEmbedUrl;
+              workItemDto.uploadedVideoUrl = portalVideoFields.uploadedVideoUrl;
+            }
```
Comment (calumjs) [inline src/backend/ipc/process-video-handlers.ts]: **Untested override (medium).** This is the change that actually fixes the #808 Tenant-view symptom (the portal payload, not the local DB), but only `derivePortalVideoFields()` is unit-tested — the wiring here is not. Nothing pins that the override fires *before* `SendWorkItemDetailsToPortal`, leaves the model's other `workItemDto` fields intact, or is skipped (model output kept) when `portalVideoFields` is `null`. The symmetric local-metadata path was extracted into a testable `applyVideoMetadataPersistence(...)` helper with wiring tests; the portal path has no equivalent. Suggest extracting `applyPortalVideoFields(dto, result)` and asserting all three cases.

Comment (calumjs) [reply src/backend/ipc/process-video-handlers.ts]: Done in 7f19588 — extracted `applyPortalVideoFields(dto, result)` (mirroring the symmetric `applyVideoMetadataPersistence` helper) and replaced the inline override in the handler with a call to it. The handler now does `applyPortalVideoFields(workItemDto, youtubeResult)` immediately before `SendWorkItemDetailsToPortal`, so the override demonstrably fires before the POST.

Added wiring tests (`applyPortalVideoFields wiring (#808 Tenant view)` in video-metadata-persistence.test.ts) asserting all three cases you called out:
- overwrites the three `uploadedVideo*` fields from the upload result;
- leaves the model's other dto fields (title/description) untouched;
- skips — leaving the model's video values in place — when the derivation returns `null` (failed upload);
- plus: clears a model-guessed embed URL to `null` for an unrecognized provider, so the Tenant view can't render a broken `<iframe>` src.

`npm run build` green, `npx vitest run --exclude '**/db/**'` 187 passed, biome clean.

File: src/backend/ipc/process-video-handlers.ts (hunk 2 above)
Comment (calumjs) [reply src/backend/ipc/process-video-handlers.ts:451]: Agreed — and you were right on the framing. The deterministic portal override (`applyPortalVideoFields` setting `uploadedVideoProvider`/`uploadedVideoEmbedUrl`/`uploadedVideoUrl` from the authoritative result before `SendWorkItemDetailsToPortal`) is now in place and is the change that resolves the Tenant-view symptom #808 reports.

I've also reframed the PR description to match reality: the portal-field derivation is now identified as the #808 fix, and the local `shaves.videoEmbedUrl` backstop is described as a separate desktop *My Shaves* thumbnail robustness improvement (a different surface that is never transmitted to the portal). This removes the false impression that the local backstop alone closes #808.

**Review by tomek-i — APPROVED** [review]
